internal/clients/clientimpl/baseimagematcher/baseimagematcher.go

```diff
@@ -0,0 +1,160 @@
+package baseimagematcher
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"slices"
+	"strings"
+
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/reporter"
+	"github.com/opencontainers/go-digest"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	maxConcurrentRequests = 1000
+)
+
+// OSVMatcher implements the VulnerabilityMatcher interface with a osv.dev client.
+// It sends out requests for every package version and does not perform caching.
+type DepsDevBaseImageMatcher struct {
+	Client   http.Client
+	Reporter reporter.Reporter
+}
+
+func (matcher *DepsDevBaseImageMatcher) MatchBaseImages(ctx context.Context, layerMetadata []models.LayerMetadata) ([][]models.BaseImageDetails, error) {
+	baseImagesMap := make([][]models.BaseImageDetails, len(layerMetadata))
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(maxConcurrentRequests)
+
+	var runningDigest digest.Digest
+	for i, l := range layerMetadata {
+		if l.DiffID == "" {
+			continue
+		}
+
+		if runningDigest == "" {
+			runningDigest = l.DiffID
+		} else {
+			runningDigest = digest.FromBytes([]byte(runningDigest + " " + l.DiffID))
+		}
+
+		chainID := runningDigest
+		g.Go(func() error {
+			if ctx.Err() != nil {
+				return ctx.Err() // this value doesn't matter to errgroup.Wait(), it will be ctx.Err()
+			}
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://api.deps.dev/v3alpha/querycontainerimages/"+chainID.String(), nil)
+			if err != nil {
```

Comment:
should we add some fault tolerance here (e.g. retries) in case of temporary service issues?

Reply:
Added the retry code from osvdev client here as well, there's quite a bit of copy and pasting, but I think that's for the best for now, since there are slight differences, and we are planning to move the osvdev code out of osvscanner, so I'm not sure how we can share code.

Reply:
Also correctly set the user agent here.